cpp/src/arrow/engine/substrait/extension_set.cc

```diff
@@ -24,6 +24,7 @@
 #include <sstream>
 #include <unordered_set>
 
+#include "arrow/compute/api.h"
```

Comment:
We try to avoid including `/api.h` files internally as they tend to be quite large and increase compilation times.  Instead we prefer many smaller includes.  The `api.h` files are for external use and also to help up distinguish public API from private API.